Allow batching the output of a join

revans2 opened the pull request:
This is the first step for out of core join. This at least partially addresses #20 

This depends on https://github.com/rapidsai/cudf/pull/8118 to go in first.

For most cases that I have tested this is strictly better than what was before. If the output of the join fits in the output batch size then the join will happen just like it does today.  If the output is larger than that we now can output it in multiple batches.  The problem that I have found is that the gather map is not spillable and after a single batch is output the GPU Semaphore is released. This means that for contrived joins that explode evenly, each active task will have a potentially large gather map in memory.  I think I can make it spillable without a lot of work. If I can then I might just do it. But I also want to spend some time running benchmarks to see if this can help fix some of the exploding join issues have have seen there.

revans2 commented on integration_tests/src/main/python/asserts.py:
```diff
@@ -28,15 +28,15 @@
                 _assert_equal(cpu[index], gpu[index], float_check, path + [index])
     elif (t is list):
-        assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {}".format(path)
+        assert len(cpu) == len(gpu), "CPU and GPU row have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
```
put list back in, it is not a row...

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -79,12 +81,669 @@ object GpuHashJoin {
+ * will be cached yet again until allowSpilling is called.
+ * When you are completely done with this object call close on it.
+ */
+trait JoinGatherer extends AutoCloseable with Arm {
```
Would it be better to move this to its own file? I would like to use it with the broadcast nested loop join/cross join code too.

abellina replied:
+1

revans2 reviewed (commented):
I'll try to add in some more java docs too.

revans2 commented:
I have been adding in spilling for the gather maps which let me push things a bit further and found a bug in the gather implementation.

https://github.com/rapidsai/cudf/issues/8121

abellina commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -79,12 +81,669 @@ object GpuHashJoin {
+        withResource(cutoff) { cutoff =>
+          withResource(cutoff.copyToHost()) { hostCutoff =>
+            hostCutoff.getInt(0)
+            hostCutoff.getInt(0)
```
It seems a little odd to have two `getInt(0)`'s here.

revans2 replied:
ctrl-d in InteliJ duplicates a line. I hit it on accident some times and do not notice, just wonder why my copy didn't work.

abellina commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -79,12 +81,669 @@ object GpuHashJoin {
+  def getBitSizeMap(n: Int): ColumnView
+
+  /**
+   * If the data is all fixed width return the size of each row, otherwise return null.
```
```suggestion
   * If the data is all fixed width return the size of each row, otherwise return None.
```

abellina commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -79,12 +81,669 @@ object GpuHashJoin {
+    data.allowSpilling()
+  }
+
+  override def getBitSizeMap(n: Int): ColumnView = synchronized {
```
Feels like this function could use a function header comment. As I understand it, it computes bit size per row of a cudf Table and returns a single column of longs with those values.

abellina commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -79,12 +81,669 @@ object GpuHashJoin {
+    val gatheredBitCount = withResource(inputBitCounts) { inputBitCounts =>
+      withResource(gatherMap.toColumnView(gatheredUpTo, n)) { gatherView =>
+        // Gather only works on a table so wrap the single column
+        val gatheredTab = withResource(new Table(inputBitCounts)) { table =>
```
nit: `getheredTable`

abellina commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -79,12 +81,669 @@ object GpuHashJoin {
+
+  override def getBitSizeMap(n: Int): ColumnView = {
+    (left.getFixedWidthBitSize, right.getFixedWidthBitSize) match {
+      case (Some(l), Some(r)) => // This should never happen, but just in case
```
should we blow up in this case if it never should happen?

abellina replied:
also, why shouldn't it happen?

abellina commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -145,249 +804,165 @@ trait GpuHashJoin extends GpuExec {
+    val finalProject: Seq[Expression] = joinType match {
+      case _: InnerLike | LeftOuter | RightOuter | FullOuter =>
+        remappedLeftOutput ++ remappedRightOutput
+//      case j: ExistenceJoin =>
```
Forgot to remove? Or need follow on work?

abellina reviewed (changes requested):
First pass @revans2

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -79,12 +81,669 @@ object GpuHashJoin {
+    }
+    if (nextCb.isEmpty) {
+      // Nothing is left to return so close ASAP.
+      close()
```
Is there a risk of a double-close error on a cudf column vector if we close here and the caller separately closes after this returns false?

revans2 replied:
Yes, but double frees are OK, although I will protect against it anyway to future proof it.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -145,249 +804,165 @@ trait GpuHashJoin extends GpuExec {
+        // We cannot dedupe anything here because the we can get nulls in the key columns
+        // at least one side
+        (left.output, left.output, right.output, right.output)
+      case _: InnerLike | LeftSemi | LeftAnti =>
```
Unlike inner join, semi and anti joins don't need anything from the right table except the keys.  I see there's special handling of these joins just below, but that's only for the final project.  It looks like we'll still cart around the right table data for these joins.

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -79,12 +81,669 @@ object GpuHashJoin {
+ */
+class JoinGathererImpl(
+    // TODO need a way to spill/cache the GatherMap
+    private val gatherMap: GatherMap,
```
Technically if you don't make it a val, except for a case class, you are accessing a closure value instead of an actual val. They might be implemented the same way, but I wanted to be explicit.

revans2 commented:
I think I have addressed all of the review comments.  I put together my own quick hack for https://github.com/rapidsai/cudf/issues/8121 and it is not enough to be able to run q72.  Even at a batch size of 26m and 200 partitions it took over 6 mins to finish one of the join tasks our of 200 and failed on the next one.  We are going to have to really think about what we want to try and do to support query 72. But all of the others run with reasonable configurations.

jlowe commented on shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/GpuBroadcastHashJoinExec.scala:
```diff
@@ -141,28 +143,18 @@ case class GpuBroadcastHashJoinExec(
+    rdd.mapPartitions { it =>
+      val builtBatch = broadcastRelation.value.batch
+      GpuColumnVector.extractBases(builtBatch).foreach(_.noWarnLeakExpected())
+      doJoin(builtBatch, it, targetSize, spillCallback,
```
Do we want to cache `targetSize` in a local `val` as is done for other parameters just above?

revans2 replied:
done

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinBase.scala:
```diff
@@ -68,37 +69,25 @@ abstract class GpuShuffledHashJoinBase(
-
-    val boundCondition = condition.map(GpuBindReferences.bindReference(_, output))
+    val spillCallback = GpuMetric.makeSpillCallback(allMetrics)
+    val localBuildOutput: Seq[Attribute] = buildPlan.output
```
Similar comment re: caching the `targetSize` in a local to help avoid serializing too much

abellina commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -79,12 +81,273 @@ object GpuHashJoin {
+  /**
+   * Filter rows from the batch where all of the keys are null.
+   */
+  def filterNulls(cb: ColumnarBatch, boundKeys: Seq[Expression]): ColumnarBatch = {
```
Should we move here the comment we had before? Sorry if I've missed it somewhere else:

> // This is a work around added in response to https://github.com/NVIDIA/spark-rapids/issues/1643.
  // to deal with slowness arising from many nulls in the build-side of the join. The work around
  // should be removed when https://github.com/rapidsai/cudf/issues/7300 is addressed.

revans2 replied:
No because it is not just for that issue any longer.  The Struct join support has extended it.

revans2 commented:
build

revans2 commented:
I upmerged and had to update the code for the new struct join support.  A good thing too because it exposed a bug in my filtering code.  It would only have been a performance regression before the struct code, but afterwards it became an error.  This should be all ready to go now. The dependency is merged.
